### PR TITLE
I18n:  Mark up explore/table for translations

### LINKED ISSUE #74312: i18n: Mark up Table for translations
### Requirements ###

Mark up https://github.com/grafana/grafana/tree/v10.1.0/public/app/features/explore/Table components for translations.

### Background

Please follow #70957 for more info and guidelines.

### PR DESCRIPTION
**Pull request by michiboo**
**What is this feature?**

Adds markup for translation in explore/table.

**Why do we need this feature?**

localization 

**Who is this feature for?**
everyone

**Which issue(s) does this PR fix?**:

Fixes #74312

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/explore/Table/TableContainer.tsx [hunk 1]
```diff
@@ -78,16 +79,16 @@ export class TableContainer extends PureComponent<Props> {
     return (
       <>
         {frames && frames.length === 0 && (
-          <PanelChrome title={'Table'} width={width} height={200}>
-            {() => <MetaInfoText metaItems={[{ value: '0 series returned' }]} />}
+          <PanelChrome title={t('table-container-title', 'Table')} width={width} height={200}>
```
Comment (ifrost) [inline public/app/features/explore/Table/TableContainer.tsx:82]: Please check [i18n guidelines](https://github.com/grafana/grafana/blob/813ebf988d354bf23e981e00c9c8458995774c21/contribute/internationalization.md#phrase-id-naming-convention). We should use 2 or 3 level message IDs depending on the context. In this case I think something like `explore.table.title` should work.

File: public/app/features/explore/Table/TableContainer.tsx [hunk 2]
```diff
@@ -78,16 +79,16 @@ export class TableContainer extends PureComponent<Props> {
...
+            {() => <MetaInfoText metaItems={[{ value: t('table-container-MetaInfoText', '0 series returned')}]} />}
```
Comment (ifrost) [inline public/app/features/explore/Table/TableContainer.tsx:83]: I think we should avoid mixing different capitalization styles and use lower case with dashes. Probably there's not need to specify the exact name of the component but use more contextual keys like, `explore.table.meta`. It'd be nice to also make use of [pluralization](https://github.com/grafana/grafana/blob/main/contribute/internationalization.md#plurals) here.

**Review by ifrost — COMMENTED** [review]
Thanks for submitting it! I left a few comments about naming convention. Please make sure you read [i18n guidelines](https://github.com/grafana/grafana/blob/main/contribute/internationalization.md).

File: public/app/features/explore/Table/TableContainer.tsx (hunk 2 above)
Comment (ifrost) [reply public/app/features/explore/Table/TableContainer.tsx:83]: Oh sorry, this message has a fixed count. Ignore my comment about pluralization.

**Comment by ifrost** [comment]
@michiboo, thanks again for contributing! Are you planning to continue to work on this one?

**Comment by michiboo** [comment]
> @michiboo, thanks again for contributing! Are you planning to continue to work on this one?

Hi yes! Thanks for the reminder, I will try finish it soon

**Comment by michiboo** [comment]
> @michiboo, thanks again for contributing! Are you planning to continue to work on this one?

Hello I just made changes based on your feedback

File: public/app/features/explore/Table/TableContainer.tsx [hunk 3]
```diff
@@ -51,9 +52,9 @@ export class TableContainer extends PureComponent<Props> {
   }
 
   getTableTitle(dataFrames: DataFrame[] | null, data: DataFrame, i: number) {
-    let title = data.name ? `Table - ${data.name}` : 'Table';
+    let title = data.name ? `${t('explore-table-title', 'Table')} - ${data.name}` : `${t('explore-table-title', 'Table')}`;
```
Comment (joshhunt) [inline public/app/features/explore/Table/TableContainer.tsx:55]: These keys do not match what's in the grafana.json files. You should avoid editing those files manually, and instead run `yarn i18n:extract` to update them. We have automated checks that will prevent merging until this is resolved.

Additionally, we shouldn't manually concatenate values into strings as this limits localisation across different languages. Instead, we can add parameters to the strings:

```ts
let title = data.name
  ? t('explore.table.title-with-name', 'Table - {{name}}', {name: data.name})
  : t('explore.table.title', 'Table');
```

File: public/app/features/explore/Table/TableContainer.tsx [hunk 4]
```diff
@@ -51,9 +52,9 @@ export class TableContainer extends PureComponent<Props> {
...
     if (dataFrames && dataFrames.length > 1) {
-      title = `Table - ${data.name || data.refId || i}`;
+      title = `${t('explore-table-title', 'Table')} - ${data.name || data.refId || i}`;
```
Comment (joshhunt) [inline public/app/features/explore/Table/TableContainer.tsx:57]: will need to be updated to match the above

File: public/app/features/explore/Table/TableContainer.tsx [hunk 5]
```diff
@@ -87,8 +88,8 @@ export class TableContainer extends PureComponent<Props> {
     return (
       <>
         {frames && frames.length === 0 && (
-          <PanelChrome title={'Table'} width={width} height={200}>
-            {() => <MetaInfoText metaItems={[{ value: '0 series returned' }]} />}
+          <PanelChrome title={t('explore-table-title', 'Table')} width={width} height={200}>
+            {() => <MetaInfoText metaItems={[{ value: t('explore-table-meta', '0 series returned')}]} />}
```
Comment (joshhunt) [inline public/app/features/explore/Table/TableContainer.tsx:92]: keys need to be updated

**Review by joshhunt — COMMENTED** [review]
This is close, but it needs keys in source should match the grafana.json files. Run `yarn i18n:extract` and commit the results

File: public/app/features/explore/Table/TableContainer.tsx (hunk 3 above)
Comment (michiboo) [reply public/app/features/explore/Table/TableContainer.tsx]: @joshhunt  thank you for the feedback, I updated according to it.

File: public/app/features/explore/Table/TableContainer.tsx [hunk 6]
```diff
@@ -87,8 +88,8 @@ export class TableContainer extends PureComponent<Props> {
     return (
       <>
         {frames && frames.length === 0 && (
-          <PanelChrome title={'Table'} width={width} height={200}>
-            {() => <MetaInfoText metaItems={[{ value: '0 series returned' }]} />}
+          <PanelChrome title={t('explore.table.title', 'Table')} width={width} height={200}>
+            {() => <MetaInfoText metaItems={[{ value: t('explore.table.meta', '0 series returned')}]} />}
```
Comment (ifrost) [inline public/app/features/explore/Table/TableContainer.tsx:92]: We can probably call it `explore.table.no-data`

**Review by ifrost — COMMENTED** [review]
Looks better now. Please make sure you sync with `main` and resolve conflicts.

File: public/app/features/explore/Table/TableContainer.tsx (hunk 6 above)
Comment (michiboo) [reply public/app/features/explore/Table/TableContainer.tsx]: @ifrost  hello I updated the name and solve the merge conflict

File: public/locales/fr-FR/grafana.json
Comment (joshhunt) [inline public/locales/fr-FR/grafana.json:1]: There are far too many changes in these files. I'm not sure how you deleted all these translations?

File: public/app/features/explore/Table/TableContainer.tsx [hunk 7]
```diff
@@ -51,9 +52,11 @@ export class TableContainer extends PureComponent<Props> {
   }
 
   getTableTitle(dataFrames: DataFrame[] | null, data: DataFrame, i: number) {
-    let title = data.name ? `Table - ${data.name}` : 'Table';
+    let title = data.name
+      ? t('explore.table.title-with-name', 'Table - {{name}}', { name: data.name })
+      : t('explore.table.title', 'Table');
     if (dataFrames && dataFrames.length > 1) {
-      title = `Table - ${data.name || data.refId || i}`;
+      title = `${t('explore.table.title-with-dataframes', 'Table')} - ${data.name || data.refId || i}`;
```
Comment (joshhunt) [inline public/app/features/explore/Table/TableContainer.tsx:59]: Like mention before, we should not manually concatenate translation strings. This should be a parameterised phrase like `explore.table.title-with-name` above it.

However, I've refactored this logic to not require the duplicate phrases.

**Comment by joshhunt** [comment]
@ifrost I refactored the `getTableTitle` function to remove the multiple calls to `t()`. Logic should be the same, but would appreciate your review, and then we can merge this is 🥳

File: public/app/features/explore/Table/TableContainer.tsx [hunk 8]
```diff
@@ -51,11 +52,12 @@ export class TableContainer extends PureComponent<Props> {
   }
 
   getTableTitle(dataFrames: DataFrame[] | null, data: DataFrame, i: number) {
-    let title = data.name ? `Table - ${data.name}` : 'Table';
-    if (dataFrames && dataFrames.length > 1) {
-      title = `Table - ${data.name || data.refId || i}`;
+    let name: string | number | undefined = data.name;
+    if (!name && (dataFrames?.length ?? 0) > 1) {
+      name = data.refId || i;
```
Comment (ifrost) [inline public/app/features/explore/Table/TableContainer.tsx:57]: Just nitpicking :D something like below would probably be more readable to me:

```suggestion
    let name = data.name;
    if (!name && (dataFrames?.length ?? 0) > 1) {
      name = data.refId || `${i}`;
```

**Review by ifrost — APPROVED** [review]
Looks good, works great 🚀

**Pull request merged**